libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/MapboxRouteLineApi.kt

```diff
@@ -373,6 +396,106 @@ class MapboxRouteLineApi(
     fun updateWithRouteProgress(routeProgress: RouteProgress) {
         updateUpcomingRoutePointIndex(routeProgress)
         updateVanishingPointState(routeProgress.currentState)
+
+        // If the de-emphasize inactive route legs feature is enabled and the vanishing route line
+        // feature is enabled and the active leg index has changed, then calling the
+        // deEmphasizeSegmentsNotInLeg() method here will get the resulting calculation cached so
+        // that calls to deEmphasizeSegmentsNotInLeg() made by updateTraveledRouteLine()
+        // won't have to wait for the result. The updateTraveledRouteLine method is much
+        // more time sensitive.
+        if (
+            routeLineOptions.deEmphasizeInactiveRouteLegs &&
+            routeLineOptions.vanishingRouteLine != null
+        ) {
+            ifNonNull(routeProgress.currentLegProgress) { routeLegProgress ->
+                if (routeLegProgress.legIndex > activeLegIndex) {
+                    jobControl.scope.launch {
+                        mutex.withLock {
+                            deEmphasizeSegmentsNotInLeg(
+                                routeLegProgress.legIndex,
+                                routeLineExpressionData
+                            )
+                            activeLegIndex = routeLegProgress.legIndex
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Adjusts the route line visibility so that only the current route leg is visible. This is
+     * intended to be used with routes that have multiple waypoints.
+     *
+     * Your activity or fragment should register a route progress listener and on each
+     * route progress call this method and render the result using the [MapboxRouteLineView].
+     *
+     * This method should NOT be used in conjunction with the vanishing route line feature. If
+     * using the vanishing route line feature ignore this method. The normal vanishing route
+     * line mechanism will take care of hiding inactive route legs if the diminish inactive
+     * route legs option is enabled AND the vanishing route line feature is enabled.
+     *
+     * @param routeProgress a [RouteProgress]
+     * @param consumer a consumer to receive the method result
+     */
+    fun showOnlyActiveLeg(
+        routeProgress: RouteProgress,
+        consumer: MapboxNavigationConsumer<Expected<RouteLineError, RouteLineUpdateValue>>
+    ) {
```

Comment:
The currently proposed API wouldn't allow us to change the styling of the route if we were not in active guidance, for example, in a preview screen.

Below I'm proposing to change the signature and avoid saying "This method should NOT be used in conjunction with the vanishing route line feature." and forcing developers into thinking how to safely use the API. We should rather document that if developers do set the option to enable different styling for inactive route legs, they should always call this method as well. This feels much more natural and decreases the risk of a mistake. If the vanishing route line is active, we should reuse the results of the last update in conjunction with the newly provided leg index to generate an update.

This simplifies the usage and we can optimize the generated results internally.

```suggestion
    fun setActiveRouteLeg(
        legIndex: Int,
        consumer: MapboxNavigationConsumer<Expected<RouteLineError, RouteLineUpdateValue>>
    ) {
```

Together with this change, we should probably also expose:
```kotlin
    fun setRoutes(
        newRoutes: List<RouteLine>,
        activeLegIndex: Int = 0,
        consumer: MapboxNavigationConsumer<Expected<RouteLineError, RouteSetValue>>
    )
```

Reply:
There are some reasons I took this approach and we should probably discuss over a call rather than trade comments here in the interest of efficiency but i'll summarize.

Your suggestion looks very similar to the approach on IOS.  However, on IOS they seem to be using a different FeatureCollection for each leg so it's easy for them to hide individual legs.  On the Android side we are using a single FeatureCollection for the whole route so calculations need to be done for the gradient line. Also we do almost all the calculations up front when the routes are set for performance reasons.  I want to avoid having to do too much recalculating during navigation and risk losing the performance gains we made.

A developer who isn't using the vanishing route line feature won't have any reason to create progress listener. At least from the point of view of the route line. They may have other reasons for doing so.  To use this alt. route leg styling feature we need the route progress but the existing updateWithRouteLegProgress() doesn't return anything to be rendered.  It didn't make sense to me to return something to be rendered in the updateWithRouteLegProgress() method because if a developer isn't using this alt. route leg styling feature there's nothing to render and since IMO most developers won't use this feature it didn't make sense to change the method signature to and document that a render method would be called that wouldn't actually do anything.

The documentation could be improved but I did a lot of thinking about the use cases and the developer experience.  Let's discuss further over a call so I can go into more detail.

Reply:
I added a method `showRouteWithLegIndexHighlighted()` to address the use case you pointed out for the route overview.  The doc. could use some polishing.